src/engine/Union.cpp

```diff
@@ -292,3 +334,185 @@ std::unique_ptr<Operation> Union::cloneImpl() const {
   }
   return copy;
 }
+
+// _____________________________________________________________________________
+std::shared_ptr<Operation> Union::createSortedVariant(
+    const vector<ColumnIndex>& sortOrder) const {
+  return std::make_shared<Union>(_executionContext, _subtrees.at(0),
+                                 _subtrees.at(1), sortOrder);
+}
+
+namespace {
+struct Wrapper {
+  const IdTable& idTable_;
+  const LocalVocab& localVocab_;
+};
+Result::IdTableVocabPair& moveOrCopy(Result::IdTableVocabPair& element) {
+  return element;
+}
+Result::IdTableVocabPair moveOrCopy(const Wrapper& element) {
+  return {element.idTable_.clone(), element.localVocab_.clone()};
+}
+}  // namespace
+
+// _____________________________________________________________________________
+// Always inline makes makes a huge difference on large datasets.
+template <size_t SPAN_SIZE>
+AD_ALWAYS_INLINE bool Union::isSmaller(const auto& row1,
+                                       const auto& row2) const {
+  using StaticRange = std::span<const ColumnIndex, SPAN_SIZE>;
+  for (auto& col : StaticRange{targetOrder_}) {
+    ColumnIndex index1 = _columnOrigins.at(col).at(0);
+    ColumnIndex index2 = _columnOrigins.at(col).at(1);
+    if (index1 == NO_COLUMN) {
+      return true;
+    }
+    if (index2 == NO_COLUMN) {
+      return false;
+    }
```

Comment:
You could remove all the sort columns after the first NO_COLUM, and precompute an
efficient return value
(If all the columns are equal, do i return true (because there was NO_COLUMN in the left input) or false (because the NO_COLUMN was right, or because all columns were equal.

Reply:
Not quite sure what exactly you want. I can't initially trim `targetOrder_`, because we still need the original values elsewhere. Otherwise this information is lost. If you're proposing adding a copy that is tailored towards the comparator, to do better loop unrolling I could certainly do that, but I don't think sort orders that just apply to a single child are common, so I'm not sure if it's worth the added complexity.

Reply:
Yes, I meant exactly something like that that in particular avoids the additional permutation of `columnOrigins.at()` as that is redundant for every line.
If that is a copy, that is no problem, as it has only to be done once for the whole input.
But If that is too much for you, then we could leave it for now, but I think we might have the opportunity for further tweaks here (especially for the common case of a single sort column, where we can statically only look at the single span of columns etc, but I think we first merge this initial version, benchmark it, then improve the query planning, and THEN we can if we have time tweak these operations.